unixfs/src/pb.rs

```diff
@@ -0,0 +1,233 @@
+use std::borrow::Cow;
+use std::convert::TryFrom;
+use std::fmt;
+use std::ops::Range;
+
+pub(crate) mod merkledag;
+pub(crate) use merkledag::PBLink;
+pub(crate) use merkledag::PBNode;
+
+pub(crate) mod unixfs;
+pub(crate) use unixfs::mod_Data::DataType as UnixFsType;
+pub(crate) use unixfs::Data as UnixFs;
+
+/// Failure cases for TryFrom conversions and most importantly, the `FlatUnixFs` conversion.
+#[derive(Debug)]
+pub enum UnixFsReadFailed {
+    /// The outer content could not be read as dag-pb PBNode.
+    InvalidDagPb(quick_protobuf::Error),
+    /// dag-pb::PBNode::Data could not be read as UnixFS::Data message
+    InvalidUnixFs(quick_protobuf::Error),
+    /// dag-pb contained zero bytes
+    NoData,
+}
+
+impl fmt::Display for UnixFsReadFailed {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use UnixFsReadFailed::*;
+        match self {
+            InvalidDagPb(e) => write!(fmt, "failed to read the block as dag-pb: {}", e),
+            InvalidUnixFs(e) => write!(
+                fmt,
+                "failed to read the dag-pb PBNode::Data as UnixFS message: {}",
+                e
+            ),
+            NoData => write!(fmt, "dag-pb PBNode::Data was missing or empty"),
+        }
+    }
+}
+
+impl std::error::Error for UnixFsReadFailed {}
+
+impl<'a> TryFrom<&'a merkledag::PBNode<'a>> for unixfs::Data<'a> {
+    type Error = UnixFsReadFailed;
+
+    fn try_from(node: &'a merkledag::PBNode<'a>) -> Result<Self, Self::Error> {
+        UnixFs::try_from(node.Data.as_ref().map(|a| a.as_ref()))
+    }
+}
+
+impl<'a> TryFrom<Option<&'a [u8]>> for unixfs::Data<'a> {
+    type Error = UnixFsReadFailed;
+
+    fn try_from(data: Option<&'a [u8]>) -> Result<Self, Self::Error> {
+        use quick_protobuf::{BytesReader, MessageRead};
+
+        let data: &'a [u8] = data.ok_or(UnixFsReadFailed::NoData)?;
```

Comment:
I think the compiler is able to infer the lifetime here?

Reply:
There is no warning from rustc or clippy here but it might be extra. I did however have to add it there at on point... I vaguely do remember wrestling this initially from `Option<Cow<'a, [u8]>>` which was a bit more involved, could be that this is from that time :)

Reply:
Can be removed, removing.